### ISSUE TITLE
[Question] Docker multi-stage builds with Bazel

### ISSUE THREAD

**Issue by corradio** [GoogleContainerTools/distroless#139]
Hi,

I'm looking for best practices to build a nodejs Docker container using Bazel.
Ideally I would use a two-stage docker build (first to build node modules + compile, second to add the compiled assets). However, I can't find how to do that using Bazel.
Is this possible?

Best,

Olivier

**Comment by mattmoor** [comment]
@dlorenc @alexeagle FYI

tl;dr There isn't a good story today, but I'd like one.

Today [rules_nodejs](https://github.com/bazelbuild/rules_nodejs) requires you to `npm install` (or equivalent) to create the `node_modules` directory in-tree, and it isn't clear to me that the way `node_binary` is set up is suitable for use by [rules_docker](https://github.com/bazelbuild/rules_docker) as a [lang_image](https://github.com/bazelbuild/rules_docker#language-rules) rule (I need to poke around the runfiles and see; I don't believe Node.js has a mono-archive form like PAR we can use as a stand-in).

In my idealized world, containerizing a Node.js application with Bazel would be as simple as rewriting `node_binary` with `node_image`, but that isn't where we're at for (at least) Node.js today.

**Comment by corradio** [comment]
Thanks for your fast answer. I was indeed looking at a similar solution.
My only worry is that by creating `node_modules` outside of the containerised environment, the native version of the modules would not be used if the blaze platform (e.g. Mac OS) is different from the container platform (e.g. Linux Ubuntu).
I wasn't able to find more documentation about the way native modules are compiled through npm in node.

Happy to contribute if we see a clear way forward.

**Comment by mattmoor** [comment]
@corradio Yes, although it is not clear to me that there is any good "cross-installation" story for any of the common package managers (e.g. `npm`, `pip`, ...), which is what we'd need to achieve this in Bazel without remote execution (where you'd actually run it on the target environment).

I think that there are two tiers of this:
1. Get this working for `node_binary` to `node_image`.
1. Get `node_binary` working for cross-compilation (a la cross-installation).

If there is a story for cross-installation in `npm` I think it makes a strong argument for moving package installation into a `repository_rule` that is sensitive to the target `Platform`.

**Comment by mattmoor** [comment]
I tried out the Dropbox `rules_node`, but hit a few issues ([1](https://github.com/dropbox/rules_node/issues/3), [2](https://github.com/dropbox/rules_node/issues/4)).

Trying out `bazelbuild/rules_nodejs` it took me a while to make sense of their samples, but I was able to get something super trivial working (as I could with Dropbox), but not with `node_modules` (see [here](https://github.com/bazelbuild/rules_nodejs/issues/60)).

**Comment by mattmoor** [comment]
So I have a basic `npm_install` rule working [here](https://github.com/bazelbuild/rules_nodejs/pull/67) with `bazelbuild/rules_nodejs` and on top of these I was able to get a `nodejs_image` rule working on top of `bazelbuild/rules_docker`, but only with `gcr.io/google-appengine/nodejs` and not `gcr.io/distroless/nodejs`.

The reason for this is that the entrypoint that `bazelbuild/rules_nodejs` creates is a bash script, so it doesn't even seem to work on `gcr.io/distroless/nodejs:debug`.

I'll keep looking at this, but that seems like good progress :)

**Comment by corradio** [comment]
Great! I'm quite busy in the next days but will try to help out as soon as I have a bit of time.

**Comment by mattmoor** [comment]
If you have a chance to check it out here's the PR: https://github.com/bazelbuild/rules_docker/pull/254

It is based on a couple PRs that need to land elsewhere (thus WiP), but I'd appreciate a second opinion on its readiness :)

thanks.

**Comment by dlorenc** [comment]
Is this still relevant?

**Comment by chanseokoh** [comment]
Closing as the question seems obsolete now.
